Fixed two critical issues of BEGINNING -> NOW streams

zzvara opened the pull request.

zzvara committed:
Typo in Client.

zzvara committed:
Fixed a case of BEGINNING -> NOW streams, where after closing the partition, a stream might gets reopened with invalid seqno.
The reopened stream then will trigger a series of RollbackExceptions and eventually a loop from which the client might not be able to recover.

themnd commented on src/main/java/com/couchbase/client/dcp/state/SessionState.java:
```diff
@@ -164,12 +165,12 @@ public void rollbackToPosition(short partition, long seqno) {
-        while (flog.hasNext()) {
-            FailoverLogEntry entry = flog.next();
+        List<FailoverLogEntry> flog = ps.getFailoverLog();
+        for (int i = 0; i < flog.size(); i ++) {
```
You should invert the cycle:

```
for (int i = flog.size() - 1; i > 0; i--) {
```

otherwise you may skip processing some items, i.e. while i is 2 for example and supposing we should remove this enrty, when we increase i to 3 in the next cycle, flog.get(3) will give you back the flog.get(4) of the original size!

zzvara replied:
@themnd Thanks!

avsej commented:
@zzvara could you, please, submit the change through gerrit? It cannot be merged using github.

zzvara commented:
@avsej I would, however http://review.couchbase.org/ stuck at loading screen for me.

I'm currently running this commit against a 4-node Spark batch job. Seems OK so far, I will submit if I find any other issues.

avsej commented:
Try to load its new UI: http://review.couchbase.org/?polygerrit=1

zzvara committed:
Added hasPassedEnd function to SessionState.

Method is useful for parallel batch reader APIs to detect that a Client has finished with all of its partitions and should be destroyed.
The hasPassedEnd is useful when in a DCP session the high seqno for any partition would overstep the end seqno.

zzvara commented:
@brett19 Is it possible for you to pull changes? Or do I just need to `git push` to the equivalent repository on Gerrit?

dnault commented:
@zzvara You need a Gerrit account in order to accept the Contributor License Agreement (CLA). Once you've accepted the agreement, the PR-bot will import the PR into Gerrit and we can review/approve it from there.

Have you tried creating a Gerrit account using the link @avsej posted? http://review.couchbase.org/?polygerrit=1

brett19 commented:
Hey @zzvara ,

There was some maintenance being performed to Gerrit periodically over the Holidays which may have been affecting your ability to access it and register.  Can you give it another try and we can go from there?

Cheers, Brett

zzvara commented:
Yes, I have registered and accepted the Agreement.

![image](https://user-images.githubusercontent.com/8371789/34495909-95ad100e-eff7-11e7-81d6-20d5bdef4b93.png)

@brett19 @dnault Is there anything I can do to ping the PR-bot to import the PR again? I've registered and with this Github user `zzvara`.

brett19 commented:
Hey @zzvara ,

Can you confirm that the email address actually used on this commit (zol******ara@gma***com) is registered to your Gerrit account, you can do that here: http://review.couchbase.org/#/settings/contact?

P.S. Sending a comment on this PR triggers the PR bot to take a new look.

Cheers, Brett

zzvara commented:
@brett19 There are no e-mail addresses listed, and after my attempt none have been registered. Reload does nothing, no e-mail received. On my profile page, there is no e-mail attach to the profile information.

Is there any way to register again?

![image](https://user-images.githubusercontent.com/8371789/34496807-4e2f8b5e-effb-11e7-885e-70b4274b4c8d.png)

avsej commented:
Probably this is because you haven't shared your email with github oauth. Just register your email you have used in commit.

brett19 commented:
Hey @zzvara,

I will investigate not receiving emails when you register a new email.  Sorry for the inconvenience.

Cheers, Brett

zzvara commented:
@brett19 Got an e-mail from Gerrit, I have been able to register my address.

I have been running a handful of Spark batch jobs lately using DCP. No issues encountered so far.

zzvara committed:
Simplified fail-over log updates in SessionState.

avsej commented:
Could you move discussion to the gerrit and close this ticket?

dnault commented:
Moving discussion to Gerrit: http://review.couchbase.org/87509

dnault commented:
Fixed in [release 0.13.0](https://github.com/couchbase/java-dcp-client/releases/tag/0.13.0).
